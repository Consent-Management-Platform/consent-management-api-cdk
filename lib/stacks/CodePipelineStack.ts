import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Effect, OpenIdConnectPrincipal, OpenIdConnectProvider, PolicyDocument, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

import { GITHUB_ORGANIZATION } from '../constants/github';
import { StageConfig } from '../interfaces/stage-config';

export interface CodePipelineStackProps extends StackProps {
  stageConfig: StageConfig;
}

/**
 * Defines infrastructure supporting automated deployment pipelines.
 */
export class CodePipelineStack extends Stack {
  constructor(scope: Construct, id: string, readonly props: CodePipelineStackProps) {
    super(scope, id, props);

    // Create GitHub OpenID Connect provider
    const oidcProvider = new OpenIdConnectProvider(this, 'GitHubOIDCProvider', {
      url: 'https://token.actions.githubusercontent.com',
      clientIds: ['sts.amazonaws.com'],
    });

    // Create GitHub principal which will be used to assume deployment roles.
    const gitHubPrincipal = new OpenIdConnectPrincipal(oidcProvider).withConditions(
      {
        StringLike: {
          'token.actions.githubusercontent.com:sub':
            `repo:${GITHUB_ORGANIZATION}/*:*`,
        },
      }
    );

    /**
      * Create a deployment role that has short lived credentials. The only
      * principal that can assume this role is the GitHub Open ID provider.
      *
      * This role is granted authority to assume AWS CDK roles.
      */
    new Role(this, 'GitHubActionsRole', {
      assumedBy: gitHubPrincipal,
      description: 'Role assumed by GitHub Actions to deploy code/infra using aws cdk',
      roleName: 'github-ci-role',
      maxSessionDuration: Duration.hours(1),
      inlinePolicies: {
        CdkDeploymentPolicy: new PolicyDocument({
          assignSids: true,
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['sts:AssumeRole'],
              resources: [`arn:aws:iam::${this.account}:role/cdk-*`],
            }),
          ],
        }),
      },
    });
  }
}
