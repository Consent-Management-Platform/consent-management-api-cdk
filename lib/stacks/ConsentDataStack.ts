import { Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, BillingMode, ProjectionType, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

import { StageConfig } from '../interfaces/stage-config';

export interface ConsentDataStackProps extends StackProps {
  stageConfig: StageConfig;
}

/**
 * Defines the consent data storage layer.
 *
 * This stack is separated from higher-level components such as API compute services
 * to minimize the risk of other layer changes impacting the consent database,
 * which should be extremely stable and never deleted after prod launch.
 */
export class ConsentDataStack extends Stack {
  public readonly consentTable: Table;

  constructor(scope: Construct, id: string, readonly props: ConsentDataStackProps) {
    super(scope, id, props);

    this.consentTable = this.createConsentTable();
    this.createConsentsByServiceUserGsi(this.consentTable);
  }

  private createConsentTable() {
    return new Table(this, 'ServiceUserConsentDynamoDBTable', {
      tableName: 'ServiceUserConsent',
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: {
        name: 'Id',
        type: AttributeType.STRING
      },
      pointInTimeRecovery: true,
      deletionProtection: true,
      billingMode: BillingMode.PAY_PER_REQUEST
    });
  }

  private createConsentsByServiceUserGsi(consentTable: Table) {
    consentTable.addGlobalSecondaryIndex({
      indexName: 'ConsentsByServiceUser',
      partitionKey: {
        name: 'UserId',
        type: AttributeType.STRING
      },
      sortKey: {
        name: 'ServiceId',
        type: AttributeType.STRING
      },
      projectionType: ProjectionType.ALL
    });
  }
}
