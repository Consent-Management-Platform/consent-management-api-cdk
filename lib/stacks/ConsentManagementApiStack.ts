import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { ApiDefinition, EndpointType, MethodLoggingLevel, SpecRestApi } from 'aws-cdk-lib/aws-apigateway';
import { AccountRootPrincipal, Effect, PolicyDocument, PolicyStatement, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { readFileSync } from 'fs';
import { join } from 'path';

import {
  CONSENT_MANAGEMENT_API_DOCS_URL,
  CONSENT_MANAGEMENT_API_ENDPOINT_EXPORT_NAME,
  CONSENT_MANAGEMENT_API_THROTTLING_BURST_LIMIT,
  CONSENT_MANAGEMENT_API_THROTTLING_LIMIT
} from '../constants/apis';
import { StageConfig } from '../interfaces/stage-config';

export interface ConsentManagementApiStackProps extends StackProps {
  apiCodePackageFilePath: string;
  stageConfig: StageConfig;
}

/**
 * Defines Consent Management API gateway and compute infrastructure.
 */
export class ConsentManagementApiStack extends Stack {
  public readonly apiLambda: Function;
  public readonly restApi: SpecRestApi;

  constructor(scope: Construct, id: string, readonly props: ConsentManagementApiStackProps) {
    super(scope, id, props);

    this.apiLambda = this.createApiLambdaFunction();
    this.restApi = this.createRestApiGateway(this.apiLambda);
  }

  private createApiLambdaFunction(): Function {
    return new Function(this, 'Consent Management API Lambda', {
      code: Code.fromAsset(this.props.apiCodePackageFilePath),
      description: 'Consent Management API Lambda',
      handler: 'com.consentframework.consentmanagement.api.ConsentManagementApiService::handleRequest',
      memorySize: 512,
      runtime: Runtime.JAVA_21,
      timeout: Duration.minutes(1)
    });
  }

  private createRestApiGateway(apiLambda: Function): SpecRestApi {
    const apiDefinition = this.constructApiDefinition(apiLambda);
    const apiAccessPolicy = this.createApiGatewayAccessPolicy();

    const api: SpecRestApi = new SpecRestApi(this, 'Consent Management API Gateway', {
      apiDefinition: ApiDefinition.fromInline(apiDefinition),
      cloudWatchRole: true,
      deployOptions: {
        metricsEnabled: true,
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: 'dev',
        throttlingBurstLimit: CONSENT_MANAGEMENT_API_THROTTLING_BURST_LIMIT,
        throttlingRateLimit: CONSENT_MANAGEMENT_API_THROTTLING_LIMIT
      },
      description: `Consent Management API, see documentation at ${CONSENT_MANAGEMENT_API_DOCS_URL}`,
      endpointExportName: CONSENT_MANAGEMENT_API_ENDPOINT_EXPORT_NAME,
      endpointTypes: [EndpointType.EDGE],
      policy: apiAccessPolicy,
      restApiName: 'ConsentManagementApi'
    });

    apiLambda.grantInvoke(new ServicePrincipal('apigateway.amazonaws.com').withConditions({
      ArnLike: { 'aws:SourceArn': api.arnForExecuteApi() }
    }));

    return api;
  }

  private constructApiDefinition(apiLambda: Function) {
    const openApiSpecFilePath = join(__dirname, '../../resources/ConsentManagementApi.openapi.json');
    const apiDefinition = JSON.parse(readFileSync(openApiSpecFilePath, 'utf8'));

    // Add API Gateway integration to OpenAPI spec
    for (const path in apiDefinition.paths) {
      for (const operation in apiDefinition.paths[path]) {
        apiDefinition.paths[path][operation]['x-amazon-apigateway-integration'] = {
          type: 'aws_proxy',
          httpMethod: 'POST',
          passthroughBehavior: 'when_no_match',
          responses: {
            default: {
              statusCode: '200',
            },
          },
          uri: `arn:aws:apigateway:${this.props.env!.region}:lambda:path/2015-03-31/functions/${apiLambda.functionArn}/invocations`
        }
      }
    }

    return apiDefinition;
  }

  /**
   * Explicitly set which principals are allowed to access which API methods.
   * Anyone not explicitly granted permissions will be denied API access.
   *
   * Ref: https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-authorization-flow.html
   */
  private createApiGatewayAccessPolicy(): PolicyDocument {
    return new PolicyDocument({
      statements: [new PolicyStatement({
        sid: 'ConsentManagementApiGatewaySameAccountFullInvokeAccess',
        actions: ['execute-api:Invoke'],
        effect: Effect.ALLOW,
        // Grant services within the same account API access
        principals: [new AccountRootPrincipal()],
        // Grants invoke permissions to all APIs within this API Gateway
        resources: ['execute-api/*']
      })]
    });
  }
}
