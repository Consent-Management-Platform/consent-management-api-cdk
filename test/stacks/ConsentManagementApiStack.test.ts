import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { join } from 'path';

import { MOCK_ENV } from '../../fixtures/mock-env';
import { MOCK_STAGE_CONFIG } from '../../fixtures/mock-stage-config';
import { ConsentManagementApiStack } from '../../lib/stacks/ConsentManagementApiStack';

describe('ConsentManagementApiStack', () => {
  it('creates the expected CloudFormation template from CDK', () => {
    const app = new App();
    const consentDataStack = new ConsentManagementApiStack(app, 'ConsentManagementApiStack', {
      env: MOCK_ENV,
      stageConfig: MOCK_STAGE_CONFIG,
      apiCodePackageFilePath: join(__dirname, '../../../consent-management-api')
    });
    const template = Template.fromStack(consentDataStack);
    expect(template.toJSON()).toMatchSnapshot();
  })
});
