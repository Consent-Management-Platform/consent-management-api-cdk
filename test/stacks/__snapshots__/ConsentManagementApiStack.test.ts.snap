// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsentManagementApiStack creates the expected CloudFormation template from CDK 1`] = `
{
  "Outputs": {},
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ConsentManagementAPIGateway2FBFF2A2": {
      "Properties": {
        "Body": {
          "components": {
            "schemas": {
              "BadRequestExceptionResponseContent": {
                "description": "This exception is thrown when the request is invalid.",
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "ConflictingResourceExceptionResponseContent": {
                "description": "This exception is thrown when a write request conflicts with stored state.",
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "Consent": {
                "description": "Record of user consent.",
                "properties": {
                  "consentData": {
                    "$ref": "#/components/schemas/ConsentData",
                  },
                  "consentId": {
                    "description": "Unique identifier for a Consent.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "consentType": {
                    "description": "Consent type.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "consentVersion": {
                    "type": "integer",
                  },
                  "expiryTime": {
                    "description": "Time when the Consent automatically expires, using the ISO 8601 format.",
                    "format": "date-time",
                    "type": "string",
                  },
                  "serviceId": {
                    "description": "Unique identifier for a service.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "status": {
                    "$ref": "#/components/schemas/ConsentStatus",
                  },
                  "userId": {
                    "description": "Unique identifier for a user.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                },
                "required": [
                  "consentId",
                  "consentVersion",
                  "serviceId",
                  "status",
                  "userId",
                ],
                "type": "object",
              },
              "ConsentData": {
                "additionalProperties": {
                  "maxLength": 256,
                  "minLength": 1,
                  "type": "string",
                },
                "description": "Consent data represented by string key-value pairs. Keys must be between 1-64 characters long, values must be between 1-256 characters long, and up to 32 key-value pairs may be provided.",
                "maxProperties": 32,
                "type": "object",
              },
              "ConsentStatus": {
                "description": "Current status of a Consent.",
                "enum": [
                  "ACTIVE",
                  "EXPIRED",
                  "REVOKED",
                ],
                "type": "string",
              },
              "CreateServiceUserConsentRequestContent": {
                "properties": {
                  "consentData": {
                    "$ref": "#/components/schemas/ConsentData",
                  },
                  "consentType": {
                    "description": "Consent type.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "expiryTime": {
                    "description": "Time when the Consent automatically expires, using the ISO 8601 format.",
                    "format": "date-time",
                    "type": "string",
                  },
                  "status": {
                    "$ref": "#/components/schemas/ConsentStatus",
                  },
                },
                "required": [
                  "status",
                ],
                "type": "object",
              },
              "CreateServiceUserConsentResponseContent": {
                "properties": {
                  "consentId": {
                    "description": "Unique identifier for a Consent.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                },
                "required": [
                  "consentId",
                ],
                "type": "object",
              },
              "GetServiceUserConsentResponseContent": {
                "properties": {
                  "data": {
                    "$ref": "#/components/schemas/Consent",
                  },
                },
                "type": "object",
              },
              "InternalServiceExceptionResponseContent": {
                "description": "This exception is thrown on an unhandled service error.",
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "ListServiceUserConsentResponseContent": {
                "properties": {
                  "data": {
                    "description": "Page of Consents",
                    "items": {
                      "$ref": "#/components/schemas/Consent",
                    },
                    "type": "array",
                  },
                  "nextPageToken": {
                    "description": "Pagination token to retrieve the next page of results.",
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "NotFoundExceptionResponseContent": {
                "description": "This exception is thrown when the requested entity is not found.",
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "UpdateServiceUserConsentRequestContent": {
                "properties": {
                  "consentData": {
                    "$ref": "#/components/schemas/ConsentData",
                  },
                  "consentType": {
                    "description": "Consent type.",
                    "pattern": "^[\\w\\-\\.]{3,128}$",
                    "type": "string",
                  },
                  "consentVersion": {
                    "description": "Updated consent version, used to conditionally write only if the stored data has the prior version, to protect against race conditions or writing stale data.",
                    "type": "integer",
                  },
                  "expiryTime": {
                    "description": "Time when the Consent automatically expires, using the ISO 8601 format.",
                    "format": "date-time",
                    "type": "string",
                  },
                  "status": {
                    "$ref": "#/components/schemas/ConsentStatus",
                  },
                },
                "required": [
                  "consentVersion",
                ],
                "type": "object",
              },
            },
          },
          "info": {
            "title": "Consent Management API",
            "version": "2024-07-03",
          },
          "openapi": "3.0.2",
          "paths": {
            "/v1/consent-management/services/{serviceId}/users/{userId}/consents": {
              "get": {
                "description": "List consents a user has granted to a given service.",
                "operationId": "ListServiceUserConsent",
                "parameters": [
                  {
                    "description": "Unique identifier for a service.",
                    "in": "path",
                    "name": "serviceId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a service.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a user.",
                    "in": "path",
                    "name": "userId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a user.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Optional. Maximum number of Consents to return in a single response.",
                    "in": "query",
                    "name": "limit",
                    "schema": {
                      "description": "Optional. Maximum number of Consents to return in a single response.",
                      "type": "integer",
                    },
                  },
                  {
                    "description": "Optional. The nextPageToken returned from the previous list request.",
                    "in": "query",
                    "name": "pageToken",
                    "schema": {
                      "description": "Optional. The nextPageToken returned from the previous list request.",
                      "type": "string",
                    },
                  },
                ],
                "responses": {
                  "200": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/ListServiceUserConsentResponseContent",
                        },
                      },
                    },
                    "description": "ListServiceUserConsent 200 response",
                  },
                  "400": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/BadRequestExceptionResponseContent",
                        },
                      },
                    },
                    "description": "BadRequestException 400 response",
                  },
                  "500": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/InternalServiceExceptionResponseContent",
                        },
                      },
                    },
                    "description": "InternalServiceException 500 response",
                  },
                },
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:apigateway:us-east-2:lambda:path/2015-03-31/functions/",
                        {
                          "Fn::GetAtt": [
                            "ConsentManagementAPILambda33B42EFA",
                            "Arn",
                          ],
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
              "post": {
                "description": "Create a user Consent for a given service.",
                "operationId": "CreateServiceUserConsent",
                "parameters": [
                  {
                    "description": "Unique identifier for a service.",
                    "in": "path",
                    "name": "serviceId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a service.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a user.",
                    "in": "path",
                    "name": "userId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a user.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                ],
                "requestBody": {
                  "content": {
                    "application/json": {
                      "schema": {
                        "$ref": "#/components/schemas/CreateServiceUserConsentRequestContent",
                      },
                    },
                  },
                  "required": true,
                },
                "responses": {
                  "200": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/CreateServiceUserConsentResponseContent",
                        },
                      },
                    },
                    "description": "CreateServiceUserConsent 200 response",
                  },
                  "400": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/BadRequestExceptionResponseContent",
                        },
                      },
                    },
                    "description": "BadRequestException 400 response",
                  },
                  "409": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/ConflictingResourceExceptionResponseContent",
                        },
                      },
                    },
                    "description": "ConflictingResourceException 409 response",
                  },
                  "500": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/InternalServiceExceptionResponseContent",
                        },
                      },
                    },
                    "description": "InternalServiceException 500 response",
                  },
                },
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:apigateway:us-east-2:lambda:path/2015-03-31/functions/",
                        {
                          "Fn::GetAtt": [
                            "ConsentManagementAPILambda33B42EFA",
                            "Arn",
                          ],
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
            },
            "/v1/consent-management/services/{serviceId}/users/{userId}/consents/{consentId}": {
              "get": {
                "description": "Retrieve a user Consent for a given service.",
                "operationId": "GetServiceUserConsent",
                "parameters": [
                  {
                    "description": "Unique identifier for a Consent.",
                    "in": "path",
                    "name": "consentId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a Consent.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a service.",
                    "in": "path",
                    "name": "serviceId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a service.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a user.",
                    "in": "path",
                    "name": "userId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a user.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                ],
                "responses": {
                  "200": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/GetServiceUserConsentResponseContent",
                        },
                      },
                    },
                    "description": "GetServiceUserConsent 200 response",
                  },
                  "404": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/NotFoundExceptionResponseContent",
                        },
                      },
                    },
                    "description": "NotFoundException 404 response",
                  },
                  "500": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/InternalServiceExceptionResponseContent",
                        },
                      },
                    },
                    "description": "InternalServiceException 500 response",
                  },
                },
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:apigateway:us-east-2:lambda:path/2015-03-31/functions/",
                        {
                          "Fn::GetAtt": [
                            "ConsentManagementAPILambda33B42EFA",
                            "Arn",
                          ],
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
              "post": {
                "description": "Update a user Consent for a given service.",
                "operationId": "UpdateServiceUserConsent",
                "parameters": [
                  {
                    "description": "Unique identifier for a Consent.",
                    "in": "path",
                    "name": "consentId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a Consent.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a service.",
                    "in": "path",
                    "name": "serviceId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a service.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                  {
                    "description": "Unique identifier for a user.",
                    "in": "path",
                    "name": "userId",
                    "required": true,
                    "schema": {
                      "description": "Unique identifier for a user.",
                      "pattern": "^[\\w\\-\\.]{3,128}$",
                      "type": "string",
                    },
                  },
                ],
                "requestBody": {
                  "content": {
                    "application/json": {
                      "schema": {
                        "$ref": "#/components/schemas/UpdateServiceUserConsentRequestContent",
                      },
                    },
                  },
                  "required": true,
                },
                "responses": {
                  "200": {
                    "description": "UpdateServiceUserConsent 200 response",
                  },
                  "400": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/BadRequestExceptionResponseContent",
                        },
                      },
                    },
                    "description": "BadRequestException 400 response",
                  },
                  "404": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/NotFoundExceptionResponseContent",
                        },
                      },
                    },
                    "description": "NotFoundException 404 response",
                  },
                  "409": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/ConflictingResourceExceptionResponseContent",
                        },
                      },
                    },
                    "description": "ConflictingResourceException 409 response",
                  },
                  "500": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "$ref": "#/components/schemas/InternalServiceExceptionResponseContent",
                        },
                      },
                    },
                    "description": "InternalServiceException 500 response",
                  },
                },
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:apigateway:us-east-2:lambda:path/2015-03-31/functions/",
                        {
                          "Fn::GetAtt": [
                            "ConsentManagementAPILambda33B42EFA",
                            "Arn",
                          ],
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
            },
          },
          "x-amazon-apigateway-request-validator": "full",
          "x-amazon-apigateway-request-validators": {
            "full": {
              "validateRequestBody": true,
              "validateRequestParameters": true,
            },
          },
        },
        "EndpointConfiguration": {
          "Types": [
            "EDGE",
          ],
        },
        "Name": "ConsentManagementApi",
        "Policy": {
          "Statement": [
            {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "execute-api/*",
              "Sid": "ConsentManagementApiGatewaySameAccountFullInvokeAccess",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ConsentManagementAPIGatewayAccount906949A1": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "ConsentManagementAPIGateway2FBFF2A2",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "ConsentManagementAPIGatewayCloudWatchRole0750B034",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "ConsentManagementAPIGatewayCloudWatchRole0750B034": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "ConsentManagementAPIGatewayDeployment": {
      "Properties": {
        "Description": "Consent Management API, see documentation at https://consent-management-platform.github.io/consent-management-api-models/v1/docs.html",
        "RestApiId": {
          "Ref": "ConsentManagementAPIGateway2FBFF2A2",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ConsentManagementAPIGatewayDeploymentStageprod": {
      "DependsOn": [
        "ConsentManagementAPIGatewayAccount906949A1",
      ],
      "Properties": {
        "MethodSettings": [
          {
            "DataTraceEnabled": false,
            "HttpMethod": "*",
            "LoggingLevel": "INFO",
            "MetricsEnabled": true,
            "ResourcePath": "/*",
            "ThrottlingBurstLimit": 5,
            "ThrottlingRateLimit": 3,
          },
        ],
        "RestApiId": {
          "Ref": "ConsentManagementAPIGateway2FBFF2A2",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ConsentManagementAPILambda33B42EFA": {
      "DependsOn": [
        "ConsentManagementAPILambdaServiceRoleDefaultPolicy85D8E30D",
        "ConsentManagementAPILambdaServiceRoleFA4B1113",
      ],
      "Properties": {
        "Description": "Consent Management API Lambda",
        "Handler": "com.consentframework.consentmanagement.api.ConsentManagementApiService::handleRequest",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "ConsentManagementApiLambdaLogGroupE3E32EB5",
          },
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "ConsentManagementAPILambdaServiceRoleFA4B1113",
            "Arn",
          ],
        },
        "Runtime": "java21",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConsentManagementAPILambdaInvoke": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ConsentManagementAPILambda33B42EFA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-2:123456789012:",
              {
                "Ref": "ConsentManagementAPIGateway2FBFF2A2",
              },
              "/*/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ConsentManagementAPILambdaServiceRoleDefaultPolicy85D8E30D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:ConditionCheckItem",
                "dynamodb:GetItem",
                "dynamodb:PutItem",
                "dynamodb:Query",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "ConsentDataStack:ExportsOutputFnGetAttServiceUserConsentDynamoDBTable7EDF0878Arn692CDB71",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "ConsentDataStack:ExportsOutputFnGetAttServiceUserConsentDynamoDBTable7EDF0878Arn692CDB71",
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
              "Sid": "ConsentDynamoDBQueryPermissions",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ConsentManagementAPILambdaServiceRoleDefaultPolicy85D8E30D",
        "Roles": [
          {
            "Ref": "ConsentManagementAPILambdaServiceRoleFA4B1113",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ConsentManagementAPILambdaServiceRoleFA4B1113": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ConsentManagementApiLambdaLogGroupE3E32EB5": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "ConsentManagementApi-Lambda-ApplicationLogs",
        "RetentionInDays": 545,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
